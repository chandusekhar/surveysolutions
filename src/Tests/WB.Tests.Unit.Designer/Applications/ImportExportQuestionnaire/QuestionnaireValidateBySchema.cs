using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Reflection;
using System.Threading.Tasks;
using ApprovalTests;
using ApprovalTests.Namers;
using ApprovalTests.Reporters;
using ApprovalTests.Reporters.TestFrameworks;
using AutoMapper;
using Main.Core.Documents;
using Newtonsoft.Json.Linq;
using Newtonsoft.Json.Schema;
using Newtonsoft.Json.Schema.Generation;
using NUnit.Framework;
using WB.Core.BoundedContexts.Designer.ImportExport;
using WB.Core.BoundedContexts.Designer.ImportExport.Models;
using WB.Infrastructure.Native.Storage;
using WB.UI.Designer.Code.ImportExport;

namespace WB.Tests.Unit.Designer.Applications.ImportExportQuestionnaire
{
    [TestFixture]
    [UseApprovalSubdirectory("QuestionnaireValidateBySchema-approved")]
    [IgnoreLineEndings(true)]
    [UseReporter(typeof(DiffReporter), typeof(NUnitReporter))]
    public class QuestionnaireValidateBySchema
    {
        [Test]
        public void when_generate_schema_should_be_equal_to_pattern_file()
        {
            JSchemaGenerator generator = new JSchemaGenerator();
            generator.SchemaReferenceHandling = SchemaReferenceHandling.All;
            generator.GenerationProviders.Add(new StringEnumGenerationProvider());
            generator.GenerationProviders.Add(new MyJSchemaGenerationProvider());
            
            JSchema jsonSchema = generator.Generate(typeof(Questionnaire));
            var json = jsonSchema.ToString();
            
            Approvals.Verify(json);
        }
        
        internal class MyJSchemaGenerationProvider : JSchemaGenerationProvider
        {
            public override JSchema GetSchema(JSchemaTypeGenerationContext context)
            {
                var schema = new JSchema();
                var descendants = typeof(IQuestionnaireEntity).Assembly.GetTypes().Where(item => !item.IsAbstract && typeof(IQuestionnaireEntity).IsAssignableFrom(item)).ToList();
                foreach (var descendant in descendants)
                {
                    // The line below never exits, because it's calling MySchemaGenerator.GetSchema again with the same parameter
                    var descendantSchema = context.Generator.Generate(descendant);
                    schema.PatternProperties.Add(descendant.Name + ".*", descendantSchema);
                }
                schema.AllowAdditionalProperties = false;
                schema.Type = JSchemaType.Object;
                return schema;
            }

            public override bool CanGenerateSchema(JSchemaTypeGenerationContext context)
            {
                return context.ObjectType == typeof(IQuestionnaireEntity);
            }
        }


        [Test]
        public async Task when_validate_questionnaire_should_be_valid_by_schema()
        {
            var questionnaireDocument = Create.QuestionnaireDocumentWithOneChapter(
                Create.TextQuestion()
            );

            var validationErrors = await Validate(questionnaireDocument);

            Assert.That(validationErrors.Count, Is.EqualTo(0));
        }

        private static async Task<ICollection<ValidationError>> Validate(QuestionnaireDocument questionnaireDocument)
        {
            var mapper = new MapperConfiguration(cfg =>
            {
                cfg.AddProfile(new QuestionnaireAutoMapperProfile());
            }).CreateMapper();

            var importExportQuestionnaireService = new ImportExportQuestionnaireMapper(mapper);
            var questionnaire = importExportQuestionnaireService.Map(questionnaireDocument);
            var json = new QuestionnaireSerializer().Serialize(questionnaire);

            var testType = typeof(QuestionnaireImportService);
            var readResourceFile = $"{testType.Namespace}.QuestionnaireSchema.json";

            await using Stream stream = testType.Assembly.GetManifestResourceStream(readResourceFile);
            using StreamReader reader = new StreamReader(stream);
            string schemaText = await reader.ReadToEndAsync();

            JSchema schema = JSchema.Parse(schemaText);

            JToken jToken = JToken.Parse(json);
            jToken.IsValid(schema, out IList<ValidationError> errors);
            return errors;
        }
    }
}