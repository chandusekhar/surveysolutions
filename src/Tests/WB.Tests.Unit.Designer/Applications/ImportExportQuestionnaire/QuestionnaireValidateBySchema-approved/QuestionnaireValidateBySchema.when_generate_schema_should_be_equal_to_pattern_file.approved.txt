{
  "definitions": {
    "Attachment": {
      "type": [
        "object",
        "null"
      ],
      "properties": {
        "Name": {
          "type": [
            "string",
            "null"
          ]
        },
        "FileName": {
          "type": [
            "string",
            "null"
          ]
        },
        "ContentType": {
          "type": [
            "string",
            "null"
          ]
        }
      },
      "required": [
        "Name",
        "FileName",
        "ContentType"
      ]
    },
    "Categories": {
      "type": [
        "object",
        "null"
      ],
      "properties": {
        "Id": {
          "type": [
            "string",
            "null"
          ]
        },
        "Name": {
          "type": [
            "string",
            "null"
          ]
        },
        "FileName": {
          "type": [
            "string",
            "null"
          ]
        }
      },
      "required": [
        "Id",
        "Name",
        "FileName"
      ]
    },
    "CoverPage": {},
    "Group": {},
    "List<Attachment>": {
      "type": [
        "array",
        "null"
      ],
      "items": {
        "$ref": "#/definitions/Attachment"
      }
    },
    "List<Categories>": {
      "type": [
        "array",
        "null"
      ],
      "items": {
        "$ref": "#/definitions/Categories"
      }
    },
    "List<Group>": {
      "type": [
        "array",
        "null"
      ],
      "items": {
        "$ref": "#/definitions/Group"
      }
    },
    "List<LookupTable>": {
      "type": [
        "array",
        "null"
      ],
      "items": {
        "$ref": "#/definitions/LookupTable"
      }
    },
    "List<Macro>": {
      "type": [
        "array",
        "null"
      ],
      "items": {
        "$ref": "#/definitions/Macro"
      }
    },
    "List<Translation>": {
      "type": [
        "array",
        "null"
      ],
      "items": {
        "$ref": "#/definitions/Translation"
      }
    },
    "LookupTable": {
      "type": [
        "object",
        "null"
      ],
      "properties": {
        "TableName": {
          "type": [
            "string",
            "null"
          ]
        },
        "FileName": {
          "type": [
            "string",
            "null"
          ]
        }
      },
      "required": [
        "TableName",
        "FileName"
      ]
    },
    "Macro": {
      "type": [
        "object",
        "null"
      ],
      "properties": {
        "Name": {
          "type": [
            "string",
            "null"
          ]
        },
        "Description": {
          "type": [
            "string",
            "null"
          ]
        },
        "Content": {
          "type": [
            "string",
            "null"
          ]
        }
      },
      "required": [
        "Name",
        "Description",
        "Content"
      ]
    },
    "QuestionnaireMetaInfo": {
      "type": [
        "object",
        "null"
      ],
      "properties": {
        "SubTitle": {
          "type": [
            "string",
            "null"
          ]
        },
        "StudyType": {
          "type": [
            "string",
            "null"
          ],
          "enum": [
            null,
            "AdministrativeRecords",
            "AgriculturalCensus",
            "AgriculturalSurvey",
            "ChildLaborSurvey",
            "CoreWelfareIndicatorsQuestionnaire",
            "DemographicAndHealthSurvey",
            "EnterpriseSurvey",
            "EnterpriseCensus",
            "InformalSectorSurvey",
            "IntegratedSurvey",
            "MultipleIndicatorClusterSurvey",
            "LaborForceSurvey",
            "LivingStandardsMeasurementStudy",
            "HouseholdHealthSurvey",
            "HouseholdSurvey",
            "PriceSurvey",
            "PrioritySurvey",
            "PopulationAndHousingCensus",
            "SampleFrame",
            "ServiceProvisionAssessments",
            "SocioEconomicMonitoringSurvey",
            "StatisticalInfoAndMonitoringProg",
            "WorldFertilitySurvey",
            "WorldHealthSurvey"
          ]
        },
        "Version": {
          "type": [
            "string",
            "null"
          ]
        },
        "VersionNotes": {
          "type": [
            "string",
            "null"
          ]
        },
        "KindOfData": {
          "type": [
            "string",
            "null"
          ]
        },
        "Country": {
          "type": [
            "string",
            "null"
          ]
        },
        "Year": {
          "type": [
            "integer",
            "null"
          ]
        },
        "Language": {
          "type": [
            "string",
            "null"
          ]
        },
        "Coverage": {
          "type": [
            "string",
            "null"
          ]
        },
        "Universe": {
          "type": [
            "string",
            "null"
          ]
        },
        "UnitOfAnalysis": {
          "type": [
            "string",
            "null"
          ]
        },
        "PrimaryInvestigator": {
          "type": [
            "string",
            "null"
          ]
        },
        "Funding": {
          "type": [
            "string",
            "null"
          ]
        },
        "Consultant": {
          "type": [
            "string",
            "null"
          ]
        },
        "ModeOfDataCollection": {
          "type": [
            "string",
            "null"
          ],
          "enum": [
            null,
            "Capi",
            "Cati",
            "FaceToFace",
            "Mail",
            "FocusGroup",
            "Internet",
            "Other"
          ]
        },
        "Notes": {
          "type": [
            "string",
            "null"
          ]
        },
        "Keywords": {
          "type": [
            "string",
            "null"
          ]
        },
        "AgreeToMakeThisQuestionnairePublic": {
          "type": "boolean"
        }
      },
      "required": [
        "SubTitle",
        "StudyType",
        "Version",
        "VersionNotes",
        "KindOfData",
        "Country",
        "Year",
        "Language",
        "Coverage",
        "Universe",
        "UnitOfAnalysis",
        "PrimaryInvestigator",
        "Funding",
        "Consultant",
        "ModeOfDataCollection",
        "Notes",
        "Keywords",
        "AgreeToMakeThisQuestionnairePublic"
      ]
    },
    "Translation": {
      "type": [
        "object",
        "null"
      ],
      "properties": {
        "Name": {
          "type": [
            "string",
            "null"
          ]
        },
        "FileName": {
          "type": [
            "string",
            "null"
          ]
        }
      },
      "required": [
        "Name",
        "FileName"
      ]
    }
  },
  "type": "object",
  "properties": {
    "Id": {
      "type": "string"
    },
    "CoverPage": {
      "$ref": "#/definitions/CoverPage"
    },
    "Children": {
      "$ref": "#/definitions/List<Group>"
    },
    "Macros": {
      "$ref": "#/definitions/List<Macro>"
    },
    "LookupTables": {
      "$ref": "#/definitions/List<LookupTable>"
    },
    "Attachments": {
      "$ref": "#/definitions/List<Attachment>"
    },
    "Translations": {
      "$ref": "#/definitions/List<Translation>"
    },
    "Categories": {
      "$ref": "#/definitions/List<Categories>"
    },
    "DefaultTranslation": {
      "type": [
        "string",
        "null"
      ]
    },
    "HideIfDisabled": {
      "type": "boolean"
    },
    "Title": {
      "type": [
        "string",
        "null"
      ]
    },
    "Description": {
      "type": [
        "string",
        "null"
      ]
    },
    "Metadata": {
      "$ref": "#/definitions/QuestionnaireMetaInfo"
    },
    "VariableName": {
      "type": [
        "string",
        "null"
      ]
    }
  },
  "required": [
    "Id",
    "CoverPage",
    "Children",
    "Macros",
    "LookupTables",
    "Attachments",
    "Translations",
    "Categories",
    "DefaultTranslation",
    "HideIfDisabled",
    "Title",
    "Description",
    "Metadata",
    "VariableName"
  ]
}